finat/aw.py

```diff
@@ -0,0 +1,158 @@
+"""Implementation of the Arnold-Winther finite elements."""
+import numpy
+
+import FIAT
+
+from gem import Literal, ListTensor
+
+from finat.fiat_elements import FiatElement
+from finat.physically_mapped import PhysicallyMappedElement, Citations
+
+
+class ArnoldWintherNC(PhysicallyMappedElement, FiatElement):
+    def __init__(self, cell, degree):
+        if Citations is not None:
+            Citations().register("Arnold2003")
+        super(ArnoldWintherNC, self).__init__(FIAT.ArnoldWintherNC(cell, degree))
+
+    def basis_transformation(self, coordinate_mapping, as_numpy=False):
+        """Note, the extra 3 dofs which are removed here
+        correspond to the constraints."""
+        V = numpy.zeros((18, 15), dtype=object)
+
+        for multiindex in numpy.ndindex(V.shape):
+            V[multiindex] = Literal(V[multiindex])
+
+        for i in range(0, 12, 2):
+            V[i, i] = Literal(1)
+
+        T = self.cell
+
+        # This bypasses the GEM wrapper.
+        that = numpy.array([T.compute_normalized_edge_tangent(i) for i in range(3)])
+        nhat = numpy.array([T.compute_normal(i) for i in range(3)])
+
+        detJ = coordinate_mapping.detJ_at([1/3, 1/3])
+        J = coordinate_mapping.jacobian_at([1/3, 1/3])
+        J_np = numpy.array([[J[0, 0], J[0, 1]],
+                            [J[1, 0], J[1, 1]]])
+        JTJ = J_np.T @ J_np
```

Comment:
The `V` matrix I'm building is a `numpy` array (this is also done in Hermite, Morley, etc) that gets converted to a `gem.ListTensor` at the end of the routine.  Is there a way to avoid this?  If I can keep everything at the `gem` level, that's great, but advice is needed.

Reply:
`ComponentTensor` basically represents a loop wrapped up as a tensor, all its entries must originate in the same rule of generation.  `ListTensor` is, on the other hand, a collection of scalar expressions that have (formally) nothing to do with each other.  Currently, there is very little in between: `Concatenate` nodes, which were introduced for a specific purpose and enjoy limited support.  Thus, it is not possible to get rid of the `ListTensor` at the end without substantial work in GEM and TSFC.

This does not mean, however, that one must use numpy up to the point of filling the `V` matrix.  Although sliced insertions won't work naïvely (e.g. `V[3:6, 3:6] = G` where `G` would be 3 x 3 GEM matrix), they are still doable, albeit a bit more complicated, for example:
```python
V[3:6, 3:6] = [[G[i, j]
                for j in range(3)]
               for i in range(3)]
```
or perhaps
```python
V[3:6, 3:6] = numpy.array([[G[i, j]
                            for j in range(3)]
                           for i in range(3)])
```
if necessary.

Reply:
Thanks, this is very helpful to my understanding.  Looking at the current code state, I'm still using numpy to do matrix manipulations (and we'll switch after the tsfc PR lands), and this will be more useful then once I've done more of the symbolic calculation at the gem rather than numpy level.